### PR TITLE
Fetching validators from POSDAO

### PR DESCRIPTION
**Pull request by saneery**
#1709

This PR for fetching validators from POSDOA and POA network.

I added:
* **Transform.Validators** - parses logs to extract new validators set
* **Fetcher.Validators** - fetches current validators from the network on the startup
* **Fetcher.ValidatorsUpdater** - periodically fetches metadata for validators in the DB (can be disabled for POS network, but necessary for POA)
* **Import.Runner.Validators** - inserts new validators and marks previous list.

For POA network required variables:
* METADATA_CONTRACT
* VALIDATORS_CONTRACT

For POS required only:
* VALIDATORS_CONTRACT

To run fetchers you need to enable them in the `indexer/config/config.exs` (for tests they already enabled):
```elixir
config :indexer, Indexer.Fetcher.Validators.Supervisor, disabled?: false
config :indexer, Indexer.Fetcher.ValidatorsUpdater.Supervisor, disabled?: false
```

### CODE REVIEW

File: apps/explorer/config/config.exs [hunk 1]
```diff
@@ -41,7 +41,7 @@ config :explorer, Explorer.Tracer,
   adapter: SpandexDatadog.Adapter,
   trace_key: :blockscout
 
-if System.get_env("METADATA_CONTRACT") && System.get_env("VALIDATORS_CONTRACT") do
+if System.get_env("VALIDATORS_CONTRACT") do
```
Comment (goodsoft) [inline apps/explorer/config/config.exs]: Not sure what's going on here, but you might want to split `config` clause in two if you wish to configure these contract addresses separately.

File: apps/explorer/test/explorer/validator/metadata_retriever_test.exs [hunk 2]
```diff
@@ -39,7 +41,20 @@ defmodule Explorer.Validator.MetadataRetrieverTest do
...
+          name: "anonymous",
+          primary: true,
+          metadata: %{
+            active: true,
+            type: "validator",
+          }
+        }
+      ]
+
+      assert MetadataRetriever.fetch_data() == expected
```
Comment (goodsoft) [inline apps/explorer/test/explorer/validator/metadata_retriever_test.exs]: Seems like you changed the essence of the test. Please update the title, if so.

File: apps/indexer/lib/indexer/transform/validators.ex [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    _ in [FunctionClauseError, MatchError] ->
+      Logger.error(fn -> "Unknown InitiateChange log format: #{inspect(log)}" end)
+      nil
+  end
+
+  defp parse_params(%{second_topic: second_topic, third_topic: nil, fourth_topic: nil} = log)
+       when not is_nil(second_topic) do
+    [addresses] = decode_data(log.data, [{:array, :address}])
+
+    MetadataRetriever.fetch_validators_metadata(addresses)
```
Comment (goodsoft) [inline apps/indexer/lib/indexer/transform/validators.ex]: Ideally, after decoding every log event (mapping `do_parse`) you should make one batch request for all validators' metadata.
Should be fine for now, if there's not a lot of `InitiateChange` events per block.

**Comment by saneery** [comment]
> You should also make a temporary worker along the lines of `UncatalogedTokenTransfers` to process the `InitiateChange` events already present in the database but not yet processed.

Thanks! I'll add it.
I have one question: how I may get validators list on the start blockscout? When there are validators in the network, but no logs about them yet?

**Comment by igorbarinov** [comment]
> > You should also make a temporary worker along the lines of `UncatalogedTokenTransfers` to process the `InitiateChange` events already present in the database but not yet processed.
> 
> Thanks! I'll add it.
> I have one question: how I may get validators list on the start blockscout? When there are validators in the network, but no logs about them yet?

from the genesis file? It can be initialized with the Staking application
@varasev please clarify

**Comment by varasev** [comment]
@saneery

> I have one question: how I may get validators list on the start blockscout? When there are validators in the network, but no logs about them yet?

You can use the [`ValidatorSet.getValidators()`](https://github.com/poanetwork/posdao-contracts/blob/f8b71990c027b5a36e07f2fe9495804d7c3d4963/contracts/abstracts/ValidatorSetBase.sol#L194) function for that. It always returns the current list of validators.

The address of `ValidatorSet` contract is `0x1000000000000000000000000000000000000001` (it is defined in `spec.json`, e.g. [like here](https://github.com/poanetwork/posdao-contracts/blob/f8b71990c027b5a36e07f2fe9495804d7c3d4963/templates/spec.json#L13)).

**Comment by saneery** [comment]
> from the genesis file? It can be initialized with the Staking application
> @varasev please clarify

No, I mean how correctly do fetching in blockscout. Maybe add an additional temporary worker to initial fetching.

**Comment by goodsoft** [comment]
You can create `fetcher/validator.ex`, which will do a one-time job of fetching current validators and shutdown normally.

**Comment by saneery** [comment]
> You can use the [`ValidatorSet.getValidators()`](https://github.com/poanetwork/posdao-contracts/blob/f8b71990c027b5a36e07f2fe9495804d7c3d4963/contracts/abstracts/ValidatorSetBase.sol#L194) function for that. It always returns the current list of validators.
> 
> The address of `ValidatorSet` contract is `0x1000000000000000000000000000000000000001` (it is defined in `spec.json`, e.g. [like here](https://github.com/poanetwork/posdao-contracts/blob/f8b71990c027b5a36e07f2fe9495804d7c3d4963/templates/spec.json#L13)).

Sure, Thanks!
This function already added
https://github.com/poanetwork/blockscout/blob/7f5405f970e5f9ae8ffe2876940b3ef43d5e090a/apps/explorer/lib/explorer/validator/metadata_retriever.ex#L22-L29

**Comment by saneery** [comment]
This PR's not correct to fetching all staking addresses and their data.
There are needs to fetch data on every block and fetch through other smart contracts methods.

**Pull request closed without merging**